File: engine/python/jsonschema_llm_engine/formatters/chat_completions.py
"""OpenAI Chat Completions API formatter."""

from __future__ import annotations

import json
from typing import Any

from jsonschema_llm_engine.exceptions import ResponseParsingError
from jsonschema_llm_engine.types import LlmRequest, ProviderConfig


def _truncate(s: str | None, max_len: int = 200) -> str:
    if s is None:
        return "<null>"
    return s if len(s) <= max_len else s[:max_len] + "..."


class ChatCompletionsFormatter:
    """OpenAI Chat Completions API formatter.

    Formats requests using ``response_format: { type: "json_schema" }``
    for structured output. Works with any endpoint that speaks the Chat
    Completions wire format (OpenAI, Azure OpenAI, etc.).

    Extracts content from ``choices[0].message.content``.
    """

    def format(
        self, prompt: str, llm_schema: Any, config: ProviderConfig
    ) -> LlmRequest:
        request_body = {
            "model": config.model,
            "messages": [{"role": "user", "content": prompt}],
            "response_format": {
                "type": "json_schema",
                "json_schema": {
                    "name": "response",
                    "strict": True,
                    "schema": llm_schema,
                },
            },
        }
        headers = {**config.headers, "Content-Type": "application/json"}
        return LlmRequest(
            url=config.url,
            headers=headers,
            body=json.dumps(request_body),
        )

    def extract_content(self, raw_response: str) -> str:
        try:
            root = json.loads(raw_response)

            choices = root.get("choices")
            if not choices or not isinstance(choices, list):
                raise ResponseParsingError(
                    f"Chat Completions response missing 'choices' array: "
                    f"{_truncate(raw_response)}"
                )

            message = choices[0].get("message")
            if message is None:
                raise ResponseParsingError(
                    f"Chat Completions response missing 'choices[0].message': "
                    f"{_truncate(raw_response)}"
                )

            content = message.get("content")
            if content is None:
                raise ResponseParsingError(
                    f"Chat Completions response has null 'content': "
                    f"{_truncate(raw_response)}"
                )

            if not isinstance(content, str):
                raise ResponseParsingError(
                    f"Chat Completions 'content' is not a string: "
                    f"{_truncate(raw_response)}"
                )

            return content
        except ResponseParsingError:
            raise
        except Exception as e:
            raise ResponseParsingError(
                f"Failed to parse Chat Completions response: {_truncate(raw_response)}"
            ) from e

File: engine/ruby/lib/json_schema_llm_engine/formatters/chat_completions.rb
# frozen_string_literal: true

# OpenAI Chat Completions formatter — formats requests for the
# /v1/chat/completions endpoint with structured output (json_schema).

require "json"

module JsonSchemaLlmEngine
  module Formatters
    class ChatCompletions
      # Format a prompt and schema into an OpenAI Chat Completions request.
      #
      # @param prompt [String] the user's natural language prompt
      # @param llm_schema [Hash] the LLM-compatible JSON Schema
      # @param config [ProviderConfig] provider endpoint/model configuration
      # @return [LlmRequest]
      def format(prompt, llm_schema, config)
        body = {
          "model" => config.model,
          "messages" => [
            { "role" => "user", "content" => prompt }
          ],
          "response_format" => {
            "type" => "json_schema",
            "json_schema" => {
              "name" => "response",
              "strict" => true,
              "schema" => llm_schema
            }
          }
        }

        headers = config.headers.merge("Content-Type" => "application/json")

        LlmRequest.new(
          url: config.url,
          headers: headers,
          body: JSON.generate(body)
        )
      end

      # Extract the generated content from an OpenAI Chat Completions response.
      #
      # @param raw_response [String] the raw response body
      # @return [String] the extracted JSON content
      # @raise [ResponseParsingError] if the response cannot be parsed
      def extract_content(raw_response)
        parsed = JSON.parse(raw_response)
        choices = parsed["choices"]

        unless choices.is_a?(Array) && !choices.empty?
          raise ResponseParsingError, "Missing or empty 'choices' in response"
        end

        content = choices.dig(0, "message", "content")

        unless content.is_a?(String) && !content.empty?
          raise ResponseParsingError, "Missing 'content' in first choice message"
        end

        content
      rescue JSON::ParserError => e
        raise ResponseParsingError, "Failed to parse response JSON: #{e.message}"
      end
    end
  end
end
